src/nvim/tui/input.c

```diff
@@ -282,27 +283,45 @@ static bool handle_focus_event(TermInput *input)
 
 static bool handle_bracketed_paste(TermInput *input)
 {
-  if (rbuffer_size(input->read_stream.buffer) > 5 &&
-      (!rbuffer_cmp(input->read_stream.buffer, "\x1b[200~", 6) ||
-       !rbuffer_cmp(input->read_stream.buffer, "\x1b[201~", 6))) {
-    bool enable = *rbuffer_get(input->read_stream.buffer, 4) == '0';
-    // Advance past the sequence
-    rbuffer_consumed(input->read_stream.buffer, 6);
-    if (input->paste_enabled == enable) {
+  RBuffer *rbuf = input->read_stream.buffer;
+
+  if (rbuffer_size(rbuf) > 5
+      && (!rbuffer_cmp(rbuf, "\x1b[200~", 6)
+          || !rbuffer_cmp(rbuf, "\x1b[201~", 6))) {
+    bool enable = *rbuffer_get(rbuf, 4) == '0';
+    rbuffer_consumed(rbuf, 6);  // Advance past the sequence
+
+    if (enable && input->paste_started) {
+      // Bogus "paste start"; forward it.
+      enqueue_input(input, "<C-v><Esc>200~", sizeof("<C-v><Esc>200~") - 1);
+      return true;
+    } else if (!enable && !input->paste_started) {
+      // Bogus "paste stop"; ignore it.
       return true;
     }
-    enqueue_input(input, PASTETOGGLE_KEY, sizeof(PASTETOGGLE_KEY) - 1);
-    input->paste_enabled = enable;
+
+    input->paste_started = enable;
+    if (enable) {
+      loop_schedule(&loop, event_create(1, apply_pastepre, 0));
+    } else {
+      flush_input(input, true);
+      enqueue_input(input, PASTEPOST_KEY, sizeof(PASTEPOST_KEY) - 1);
+    }
     return true;
   }
   return false;
 }
 
+static void apply_pastepre(void **argv)  // MAIN thread
+{
+  apply_autocmds(EVENT_PASTEPRE, NULL, NULL, false, curbuf);
```

Comment:
When I asked about allowed I did not mean *technical* reasons not to call `apply_autocmds`. I talked about architectural issues, even if you call `vim_doautocmd(STATIC_CSTR_AS_STRING("PastePre"), STATIC_CSTR_AS_STRING("") /* group */, STATIC_CSTR_AS_STRING("") /* fname */, BOOLEAN_OBJ(false).data.boolean /* nomodeline */, &err)` (note I did not suggest using `vim_command` because I do not like the idea; also in any case every command functionality will eventually use API) this still needs to be in the main thread because here you are using a C function avoiding events dispatcher.